Battle monster attack

Hae1385 opened the pull request.

Hae1385 committed:
✨ Add: BattleMonsterAttack.cs 추가

Hae1385 committed:
Delete TextRPG_Team25/BattleMonsterAttack.cs

Hae1385 committed:
✨ Add: 몬스터 공격 메소드 생성

- 생성된 적 개체수 만큼 오차값 10퍼센트 확률로 플레이어 피해

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,42 @@
     {
+        public void MonsterAttack()
+        {
+            int battleHp = player.hp; //전투에서쓰는 HP따로 선언
```
현재 함수 내에서만 변동이 있도록 설정하셨는데, 이렇게 하면 플레이어의 실제 체력에 변함이 없기 때문에 적 턴이 끝나고 내 턴으로 넘어왔을 때 체력 변화를 줄 수 없습니다! 또 배틀이 끝났을 때 체력을 회복하기 위한 포션을 마신다든지 하는 행동을 할 수 없게 됩니다.

대신 현재 체력은 저장해놓고 (변화량을 보여줘야하기 때문에) 내부에서는 직접적으로 player.hp -= damage 를 조정하는 방법이 좋을 것 같아요. 그렇게 되면 현재 함수가 실행된 후 플레이어 공격 함수가 실행됐을 때도 플레이어의 hp를 공유할 수 있겠죠?

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,42 @@
+            int battleHp = player.hp; //전투에서쓰는 HP따로 선언
+
+
+        for (int battleEnermy = 0; battleEnermy <= EnermyList.Count; battleEnermy++)  //랜덤으로 생성된 적 개체수만큼 실행
```
Enemy 스펠링 오타 있습니다!
그리고 적은 현재 monster 용어를 사용하고 있기 때문에 적 지칭은 monster로 통일해주면 보기 편할 것 같아요

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,42 @@
+        for (int battleEnermy = 0; battleEnermy <= EnermyList.Count; battleEnermy++)  //랜덤으로 생성된 적 개체수만큼 실행
+            {
+
+                int randomDmg = new Random().Next(1, 11) //적 공격력의 오차값 10퍼센트
```
적 공격력의 10%가 오차값이기 때문에 따로 설정해야 합니다..! 제가 플레이어 공격 구현할 때에는 아래와 같이 구현했습니다
```suggestion
                int offset = (int)Math.Ceiling(baseAttack * 0.1f); // Math.Celiling: 올림 함수
                int damage = new Random().Next(baseAttack - offset, baseAttack + offset + 1);
```

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,42 @@
+                        Console.WriteLine($"HP {battleHP} -> {battleHP - enermyDmg}") // 플레이어가 받은피해와 플레이어 체력 출력
+
+
+                    battleHP -= enermyDmg //위의 랜덤공격력으로 플레이어 HP감소
```
```suggestion
                    player.hp -= enermyDmg //위의 랜덤공격력으로 플레이어 HP감소
```

위에서 코멘트한 내용입니당

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,42 @@
+
+                        if (EnermyList[i].EnermyDead == true) { randomDmg = 0 }
+
+                        Console.WriteLine($"Lv.{Monster.level} {Monster.name} 의 공격!");
```
몬스터는 정적 클래스가 아니기 때문에 바로 사용할 수 없습니다! 따라서 객체를 만든 후에 사용하셔야 해요. 현재로서는 어떤 몬스터를 때릴지 지정이 안 되어있는 상태입니다. 그렇기 때문에 더더욱 몬스터 공격을 Monster 클래스 내부의 메서드로 옮겨야겠네요..! 

```suggestion
                        Console.WriteLine($"Lv.{level} {name} 의 공격!");
```

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
✨ Add.설명추가

설명추가

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
Update Battle.cs

Hae1385 committed:
🐛 Fix: 버그 수정

Hae1385 committed:
🐛 Fix: 코드 전면 수정

몬스터의 공격의 10퍼센트 확률로 늘거나 줄어드는 것이 아닌 동일확률로 때리게 변경했습니다. 플레이어의 HP가 코드 안에서만 줄어드는 것이 아닌 실제 플레이어의 HP가 감소하도록 변경하였습니다. 불필요한 메소드를 삭제하였습니다.

Hae1385 committed:
🐛 Fix: 오타 수정

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,41 @@
+                Console.WriteLine($"{player.name} 을(를) 맞췄습니다. [데미지 : {damage}]");
+                Console.WriteLine();
+                Console.WriteLine($"Lv.{player.level} {player.name}");
+                Console.WriteLine($"HP {player.hp} -> {Math.Max(player.hp - damage, 0)}"); //Math.Max로 더 큰 값 출력 player의 HP가 0아래로 떨어졌을때 0을 띄우기 위함
```
체크 좋아요!! 실제로 값과 상관없이 0을 띄우려면 이 방법이 맞습니다!

다만 현재 로직에서는 실제 플레이어 체력도 0 아래로 내려가면 안되기 때문에 데미지 들어간 후 
player.hp = 0; 이렇게 수정해주시면 됩니다! 작성하신 코드 아래에 if (player.hp == 0) 이부분에서는 player 체력이 0인지 확인하기도 하니까요!

* 코드는 위에서부터 순서대로 실행되기 때문에 플레이어 실제 체력을 깎은 후 안내 메세지가 나오는 편이 좋을 것 같습니다. 현재 메서드에서 로그 출력 코드는 가장 아래로 내려주세요!

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,41 @@
+                Console.WriteLine();
+
+                //체력 감소
+                player.hp -= damage;
```
이겁니다!!!!!! 이렇게 해야 플레이어 체력이 실제로 깎여요 굳 👍🏻

happy124210 commented on TextRPG_Team25/Battle.cs:
```diff
@@ -1,7 +1,41 @@
-﻿namespace TextRPG_Team25
+﻿using System.ComponentModel.Design;
+using System.Numerics;
+using System.Security.Cryptography.X509Certificates;
```
가끔가다가 코드 자동완성이 잘못 되면 이런 이상한 using문이 자동으로 추가될 때가 있습니다 ㅋ큐ㅠㅠ 커밋할 땐 필요없는 using문은 전부 지운 후 보내주세요! 사용하지 않는 using문은 회색으로 비활성화 되어있을 거예요!

happy124210 reviewed (changes requested):
해성님 코드는 딱 보면 왜 이렇게 짰는지 이해가 가서 좋습니당 저도 별건 아니지만 많이 도와드릴테니 계속 이렇게 해주시면 금방 같이 성장할 수 있을 것 같아요!!!! 감사합니당 수고하셨어요 👏🏻👏🏻
